File: frontend/src/api/suites.ts
import request from '@/utils/request'

import type { VariableChip } from '@/utils/caseVariables'
import type { Step } from './cases'

export interface Suite {
  id: number
  project_id: number
  name: string
  description?: string | null
  status: string
  /** 套件模块树（scope='suite'）的模块；null 表示未分组 */
  module_id?: number | null
  module_name?: string | null
  created_by?: number | null
  case_count: number
  // 方案 §2：套件前后置步骤（Action Step，与用例步骤同构）
  setup_steps?: Step[]
  teardown_steps?: Step[]
  created_at: string
  updated_at: string
}

/** 行内变量摘要（最多 2 项），与共享展示口径一致。 */
export type SuiteVariablePreview = VariableChip

export interface SuiteCase {
  id: number
  case_id: number
  case_name: string
  module_name?: string | null
  sort_order: number
  variable_count: number
  variables_preview: SuiteVariablePreview[]
}

export interface SuiteCaseVariable {
  name: string
  reference_count: number
  status: string
  display_value: string
  inherited_value: string | null
  inherited_scope: string | null
  override_enabled: boolean
  override_value: string
}

export interface SuiteCaseVariables {
  suite_id: number
  membership_id: number
  case_id: number
  case_name: string
  test_asset_revision: number
  total: number
  variables: SuiteCaseVariable[]
}

export interface Variable {
  id: number
  scope: string
  project_id?: number | null
  suite_id?: number | null
  case_id?: number | null
  name: string
  /** 敏感变量接口返回掩码；编辑时不得将该值回写。 */
  value: string
  kind: 'fixed' | 'random_integer' | 'random_choice'
  spec?: { min?: number; max?: number; items?: string[] } | null
  description?: string | null
  is_sensitive: boolean
  created_at: string
  updated_at: string
}

export interface SuitePage {
  total: number
  page: number
  page_size: number
  items: Suite[]
}

export const VARIABLE_SCOPES = [
  { value: 'global', label: '全局' },
  { value: 'project', label: '项目' },
  { value: 'suite', label: '套件' },
  { value: 'case', label: '用例' },
]

export interface SuiteListParams {
  page?: number
  page_size?: number
  keyword?: string
  status?: string
  /** 具体模块：后端会包含其子孙模块 */
  module_id?: number
  /** 未分组（`module_id IS NULL`）。**不要**用 `module_id: null` 表达，axios 会丢弃 null 参数 */
  ungrouped?: boolean
}

export async function listSuites(projectId: number, params: SuiteListParams = {}) {
  return request.get<SuitePage>(`/projects/${projectId}/suites`, { params })
}

export function createSuite(projectId: number, data: {
  name: string
  description?: string
  module_id?: number | null
  setup_steps?: Step[]
  teardown_steps?: Step[]
}) {
  return request.post<Suite>(`/projects/${projectId}/suites`, data)
}

export function getSuite(id: number) {
  return request.get<Suite>(`/suites/${id}`)
}

export function updateSuite(id: number, data: Partial<Suite>) {
  return request.put<Suite>(`/suites/${id}`, data)
}

export function deleteSuite(id: number) {
  return request.delete<void>(`/suites/${id}`)
}

export function listSuiteCases(suiteId: number) {
  return request.get<SuiteCase[]>(`/suites/${suiteId}/cases`)
}

export function getSuiteCaseVariables(suiteId: number, membershipId: number) {
  return request.get<SuiteCaseVariables>(`/suites/${suiteId}/cases/${membershipId}/variables`)
}

/** 增量覆盖：字符串=设置编排项覆盖，null=删除覆盖并恢复继承。 */
export function patchSuiteCaseVariables(
  suiteId: number,
  membershipId: number,
  updates: Record<string, string | null>,
) {
  return request.patch<SuiteCaseVariables>(
    `/suites/${suiteId}/cases/${membershipId}/variable-overrides`,
    { updates },
  )
}

export function addSuiteCases(suiteId: number, caseIds: number[]) {
  return request.post<SuiteCase[]>(`/suites/${suiteId}/cases`, { case_ids: caseIds })
}

export function reorderSuiteCases(suiteId: number, membershipIds: number[]) {
  return request.put<void>(`/suites/${suiteId}/cases/order`, { membership_ids: membershipIds })
}

export function removeSuiteCase(suiteId: number, membershipId: number) {
  return request.delete<void>(`/suites/${suiteId}/cases/${membershipId}`)
}

export function listVariables(params?: {
  scope?: string
  project_id?: number
  suite_id?: number
  case_id?: number
}) {
  return request.get<Variable[]>('/variables', { params })
}

export function createVariable(data: Partial<Variable>) {
  return request.post<Variable>('/variables', data)
}

export function updateVariable(id: number, data: { value?: string; kind?: Variable['kind']; spec?: Variable['spec']; description?: string | null; is_sensitive?: boolean }) {
  return request.put<Variable>(`/variables/${id}`, data)
}

export function deleteVariable(id: number) {
  return request.delete<void>(`/variables/${id}`)
}

File: backend/app/api/suites.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_editable_project, get_project_permission
from app.core.database import get_db
from app.core.errors import api_error
from app.models import Project, User
from app.schemas.suite import (
    SuiteAddCaseRequest,
    SuiteCaseOut,
    SuiteCaseVariableOverrideRequest,
    SuiteCaseVariablesOut,
    SuiteCreate,
    SuiteOut,
    SuitePage,
    SuiteReorderRequest,
    SuiteUpdate,
)
from app.services import suite_service
from app.services.profile_resolver_nodes import ProfileRuleError
from app.utils.pagination import get_pagination

router = APIRouter(tags=["套件管理"])


def _suite_out(suite, case_count: int, module_name: str | None = None) -> SuiteOut:
    result = SuiteOut.model_validate(suite)
    result.case_count = case_count
    result.module_name = module_name
    return result


@router.get("/projects/{project_id}/suites", response_model=SuitePage)
async def list_suites(
    project_id: int,
    pagination=Depends(get_pagination),
    module_id: int | None = None,
    ungrouped: bool = False,
    keyword: str = "",
    status_filter: str = Query(default="", alias="status"),
    _perm: tuple[Project, str | None] = Depends(get_project_permission),
    db: AsyncSession = Depends(get_db),
):
    total, suites, counts, module_names = await suite_service.list_page(
        db,
        project_id=project_id,
        module_id=module_id,
        ungrouped=ungrouped,
        keyword=keyword,
        status=status_filter,
        offset=pagination.offset,
        limit=pagination.limit,
    )
    return {
        "total": total,
        "page": pagination.page,
        "page_size": pagination.page_size,
        "items": [
            _suite_out(
                suite,
                counts.get(suite.id, 0),
                module_names.get(suite.module_id) if suite.module_id else None,
            )
            for suite in suites
        ],
    }


@router.post("/projects/{project_id}/suites", response_model=SuiteOut, status_code=status.HTTP_201_CREATED)
async def create_suite(
    project_id: int,
    body: SuiteCreate,
    project: Project = Depends(get_editable_project),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await suite_service.create(db, project_id=project_id, body=body, user_id=user.id)


@router.get("/suites/{suite_id}", response_model=SuiteOut)
async def get_suite(
    suite_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    await get_project_permission(suite.project_id, user, db)
    return _suite_out(
        suite,
        await suite_service.count_cases(db, suite.id),
        await suite_service.module_name_of(db, suite),
    )


@router.put("/suites/{suite_id}", response_model=SuiteOut)
async def update_suite(
    suite_id: int,
    body: SuiteUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    return await suite_service.update(db, suite=suite, body=body)


@router.delete("/suites/{suite_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_suite(
    suite_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    await suite_service.delete(db, suite=suite)


@router.get("/suites/{suite_id}/cases", response_model=list[SuiteCaseOut])
async def list_suite_cases(
    suite_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    await get_project_permission(suite.project_id, user, db)
    return [
        SuiteCaseOut(**item)
        for item in await suite_service.list_cases_with_variables(db, suite)
    ]


@router.get(
    "/suites/{suite_id}/cases/{membership_id}/variables",
    response_model=SuiteCaseVariablesOut,
)
async def get_suite_case_variables(
    suite_id: int,
    membership_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    await get_project_permission(suite.project_id, user, db)
    return await suite_service.membership_variables(
        db, suite=suite, membership_id=membership_id
    )


@router.patch(
    "/suites/{suite_id}/cases/{membership_id}/variable-overrides",
    response_model=SuiteCaseVariablesOut,
)
async def patch_suite_case_variables(
    suite_id: int,
    membership_id: int,
    body: SuiteCaseVariableOverrideRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    try:
        return await suite_service.update_membership_variables(
            db, suite=suite, membership_id=membership_id,
            updates=body.updates, user_id=user.id,
        )
    except ProfileRuleError as exc:
        raise api_error(
            status.HTTP_422_UNPROCESSABLE_CONTENT, exc.code, exc.message
        ) from None


@router.post(
    "/suites/{suite_id}/cases",
    response_model=list[SuiteCaseOut],
    status_code=status.HTTP_201_CREATED,
)
async def add_suite_case(
    suite_id: int,
    body: SuiteAddCaseRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    created = await suite_service.add_cases(db, suite=suite, case_ids=body.case_ids or [])
    created_ids = {relation.id for relation, _case in created}
    items = await suite_service.list_cases_with_variables(db, suite)
    return [SuiteCaseOut(**item) for item in items if item["id"] in created_ids]


@router.put("/suites/{suite_id}/cases/order", status_code=status.HTTP_204_NO_CONTENT)
async def reorder_suite_cases(
    suite_id: int,
    body: SuiteReorderRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    await suite_service.reorder_cases(db, suite=suite, membership_ids=body.membership_ids)


@router.delete("/suites/{suite_id}/cases/{membership_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_suite_case(
    suite_id: int,
    membership_id: int,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    suite = await suite_service.get_or_404(db, suite_id)
    _project, role = await get_project_permission(suite.project_id, user, db)
    if role not in ("owner", "admin", "member"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="权限不足")
    await suite_service.remove_case(db, suite=suite, membership_id=membership_id)
